### ISSUE TITLE
lib64/libcrypto.so.3: version `OPENSSL_3.0.3' not found

### ISSUE THREAD

**Issue by cosmobille** [openssl/openssl#19801]
Fedora 36, OpenSSL v3.0.3 new install. 
 $ openssl help
openssl: /lib64/libcrypto.so.3: version `OPENSSL_3.0.3' not found (required by openssl)
$ pwd
/usr/local/lib64
 $ ls
bpf        libcrypto.a   libcrypto.so.3  libssl.so    ossl-modules  pkgconfig
engines-3  libcrypto.so  libssl.a        libssl.so.3  perl5

**Comment by tmshort** [comment]
Do you have two openssl installs? It appears that you have the Fedora OpenSSL libraries in /lib and a local install in /usr/local/lib64. _The linux loader won't pick up libraries from the current directory._

You don't indicate which openssl is in use: `which openssl`. If you have `/usr/local/bin` before `/usr/bin` in your `${PATH}`, then you are picking up the local install binary, but it's finding the system installed libraries in `/lib64`. If you want to use the local installed libraries, you may need to update `ldconfig` and/or set the `LD_LIBRARY_PATH=/usr/local/lib64`.
e.g.
```
LD_LIBRARY_PATH=/usr/local/lib64 /usr/local/bin/openssl help
```
vs.
```
/usr/bin/openssl help
```

**Comment by cosmobille** [comment]
Thanks @tmshort. I installed OpenSSL 3.0.3. I don't see an openssl directory in /lib64. I did have OpenSSL 1.1.1 installed as a component of Anaconda 3. I moved it temporarily but cannot uninstall it as it appears to be necessary for essentially all other software in Anaconda, probably for all installs & upgrades. 

I did a system update from Fedora 36 to 37 to see if that would work but it didn't seem to change anything. I don't see that Fedora 37 installed it's own version. 

I set a path to LD_LIBRARY_PATH in ~/.bashrc as you suggested. That did work as long as the version in Anaconda anaconda3/bin/openssl is removed. As soon as I replace openssl binary back in anaconda3/bin, it executes that version & ignores the v3 install in /usr/local. I did set the path for OpenSSL previous to the path to Anaconda in .bashrc, reloaded $ source ~/.bashrc but still it executes v1.1.1 from Anaconda. As soon as I remove that version & reload the source, it executes v3.  

My ~/.bashrc contains:
# PATH setting for OpenSSL
export PATH="/usr/local:$PATH"
export LD_LIBRARY_PATH="/usr/local/lib64:$PATH"

I'm trying to configure Apache & need OpenSSL 3.

**Comment by cosmobille** [comment]
Oops.

< 
#PATH setting for OpenSSL
export PATH="/usr/local:$PATH"
export LD_LIBRARY_PATH="/usr/local/lib64:$PATH" 
>

**Comment by tmshort** [comment]
So, you should only ever have one `openssl` executable. You _can_ have multiple versions of the library files (i.e. libcrypto and libssl) as they are versioned. **But**, you should also only have one instance of each significant version. i.e. Don't install two versions of OpenSSL 3.0, as they share library names. You also can only have one set of include files, as they are not versioned.

Assuming you installed from source, I would suggest removing that. i.e. remove the version in `/usr/local`

Fedora 36 and 37 require OpenSSL 3.0, as many of their system utilities (e.g. grub, sshd) rely on that version.

Since you are on Fedora, use `dnf` to install `openssl` and `openssl1.1` and that way you'll have both versions of the library. The problem is you can't have both "devel" versions, because the header files conflict (the `dnf` command errors out). You can always swap between the two (i.e. uninstall `openssl-devel` to install `openssl1.1-devel`).

I will note that Fedora includes a version of anaconda that can be installed via `dnf`. It's a python script, so it relies on the system python.

By using the Fedora provided libraries and `dnf`, you can ensure that you won't have conflicts.

If you insist on installing your own version of OpenSSL:

1. Use different prefixes (and other configuration options), put them somewhere in `/usr/local`
2. Use any `--with-openssl`-type options when configuring/building components, use the `-rpath` (or equivalent option) when linking to specify a non-standard library path, or link statically
3. Don't install the OpenSSL executable; use the system `openssl` instead. Or at least, don't put the locally-build version location in your PATH.
4. The LD_LIBRARY_PATH variable should only be set when running an application that needs the locally installed version. Don't set it in your `.bashrc`. Oftentimes it's set in a script that then runs the program in question:
```
#!/bin/sh
# openssl11 script
LD_LIBRARY_PATH=/usr/local/lib64 /usr/local/bin/openssl "$@"
```

Basically, you need to keep the version of OpenSSL you install quite separate from the rest of your system, and use it only for the specific application that needs it. Don't use it yourself, use the system OpenSSL where possible.

**Comment by beldmit** [comment]
@t8m we probably need this text somewhere not only here

**Comment by t8m** [comment]
I do not think Fedora-specific text should be part of upstream documentation. If you can think about something general in regards to installing OpenSSL from source on Linux distributions, it would be welcome.

**Comment by tmshort** [comment]
This generally applies to all software: only install one version!

There are some exceptions, and that's only because the software is designed for it (e.g. python2 vs python3). 

My answer is specific to Fedora, since I went through a recent Fedora 36/37 install, so it's very top-of-mind. But the answer could apply to any Linux distribution (macOS and Windows are totally different beasts).

**Comment by cosmobille** [comment]
@tmshort 
How do I remove only the version I installed from source? Do I need to go through /usr/local & just try to find all the pieces? I don't see an uninstall command in the install docs & don't want to do it from dnf. 
This is what I have installed. I commented out path to Anaconda3 in .bashrc & moved it into another directory. Also commented out previous paths to OpenSSL I'd added. 
This is what I apparently have left. I don't know what I'm looking at. Trying to safely remove only the one I installed in /usr/local. 

<$ rpm -qa | grep openssl
openssl1.1-1.1.1q-2.fc37.x86_64
openssl-pkcs11-0.4.12-2.fc37.x86_64
openssl-libs-3.0.5-3.fc37.x86_64
apr-util-openssl-1.6.1-22.fc37.x86_64
openssl-3.0.5-3.fc37.x86_64
openssl-devel-3.0.5-3.fc37.x86_64
xmlsec1-openssl-1.2.34-4.fc37.x86_64>

Why does Fedora 37 use the old version of OpenSSL if it's been updated? I thought it was necessary to use v3 for Apache. The reason I installed OpeSSL in the first place was because I installed Apache from their site & couldn't get it run, problem with a certificate, which I thought was due to no OpenSSL. Oops. 

thanks

**Comment by beldmit** [comment]
@tmshort I'm sorry, I improperly read the author of the text.

I think it should go to OpenSSL Wiki or to a separate README. I also think it's quite universal (possibly after small tweaking) and not Fedora-specific.

**Comment by tmshort** [comment]
> How do I remove only the version I installed from source? Do I need to go through /usr/local & just try to find all the pieces? I don't see an uninstall command in the install docs & don't want to do it from dnf.

Assuming you did `make install`, you can do `make uninstall` within the OpenSSL source directory. Otherwise, clean out the `/usr/local` directory of OpenSSL stuff. If you defined `DESTDIR` when doing `make install`, then you will also need to include that during `make uninstall`. If you specified any prefixes, openssldirs, etc (you can discover these locations via `openssl version -a`), you may need to clean out those locations as well.

> Why does Fedora 37 use the old version of OpenSSL if it's been updated?

Fedora 37 uses OpenSSL 3.0.5, which is a fairly recent version, but also backport fixes. Many distributions will stick with a base version (e.g. 3.0.5 here) and backport fixes so as to keep the system stable. For example, the project withdrew 3.0.6, so if Fedora had moved to 3.0.6 before it was withdrawn, Fedora would've been in a pickle.

Fedora makes OpenSSL 1.1.1q _available_ as user-installed package, it's not the default OpenSSL. This is to handle applications that were built with OpenSSL 1.1.1, and have not updated to 3.0.

For example:
```
tmshort@cube:~$ ldd `which sshd` | grep libcrypto
        libcrypto.so.3 => /lib64/libcrypto.so.3 (0x00007f64f9600000)
tmshort@cube:~$ ldd /etc/httpd/modules/mod_ssl.so | grep libssl
        libssl.so.3 => /lib64/libssl.so.3 (0x00007fa43d2f2000)
tshort@cube:~$ dnf repoquery --requires mod_ssl | grep OPENSSL
Last metadata expiration check: 0:00:43 ago on Thu 08 Dec 2022 01:26:12 PM EST.
libcrypto.so.3(OPENSSL_3.0.0)(64bit)
libssl.so.3(OPENSSL_3.0.0)(64bit)
tshort@cube:~$
```

As a side note, Linux uses SONAMES (e.g. `objdump -p /lib64/libssl.so`), and function versions (e.g. `objdump -t /lb64/libssl.so`) to correctly identify and load the correct dynamic library.

But to answer the question I think you meant:

The thing is, _you haven't updated it_. You installed it in `/usr/local/lib` rather than installing it into `/lib64`, Linux uses `ldconfig` to manage where to load dynamic libraries, and I'm assuming you didn't add configuration for `/usr/local/lib` to `ldconfig` nor run the command to update the cache.

You were trying to run a new version of the `openssl` executable without specifying where the new libraries were.

If you really want to update a systems OpenSSL, there are a few things to do. However, _I don't recommend_ doing this, and instead, it's better to pull OpenSSL down from your distribution's repos:

## Not recommended:

0. Build OpenSSL with options compatible with your system (Fedora in this case). This might require downloading the RPM or DEB file, extracting it, and examining how `Configure`/`config` is invoked. 
1. You may want to install OpenSSL into a temporary directory, by specifying the `DESTDIR` variable. This will make finding the files to copy in the following step easier.
2. Run the existing openssl command to figure out where stuff is located:
```
ldd openssl
openssl version -a
```
3. Copy the `libssl.so.v.e.r` and `libcrypto.so.v.e.r` files into /lib64 (or wherever they are located in your distribution). In this case, `v.e.r` might be `3.0.7` (the most recent).
4. Delete the symlinks `libcrypto.so` and `libssl.so` and recreate them, pointing to `libssl.so.v.e.r` and `libcrypto.so.v.e.r`.  You may also need re-symlink`libssl.so.3` and `libcrypto.so.3` to the new versions. (These two steps will keep `libssl.so.3.0.5` and `libcrypto.so.3.0.5` around in case any library is looking for them directly.). 
5. Rename the existing `openssl` executable (e.g. /usr/bin/openssl) to something like `openssl-sys`, and then copy the newly compiled `openssl` into that directory.
6. Update the providers located in `/usr/lib64/ossl-modules` with the modules built in your local build (just replace them... maybe... replacing `fips.so` will likely run afoul of FIPS requirements).
7. Update the ENGINEs located in `/usr/lib64/engines-3` (there may also be an `engines-1.1` directory, which applies only for 1.1.x versions).
8. Update the `/etc/pki/tls` directory, which includes certificates and configuration files. If `fipsmodule.cnf` is not updated  when `fips.so` is updated, you will get a failure when trying to use the `fips` provider.

Another method would be to create a new RPM or DEB file, but that is a bit beyond the scope of this comment.

**Comment by cosmobille** [comment]
I got it. I just pulled all the pieces out of the various dictories in /usr/local. It executes the Fedora 3.0.5 version now. Thankyou everybody for all your help.Much appreciated.
